### PR TITLE
use a single binary

### PR DESCRIPTION
**Pull request by shan4993**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -65,7 +65,7 @@ jobs:
     steps:
       - run: 'echo prechecks complete'
 
-  # Build github-metrics-aggregator-webhook and push to artifact registry
+  # Build gmactl and push to artifact registry
```
Comment (sethvargo) [inline .github/workflows/ci.yml:68]: Nit: I dislike "ctl" commands. Can we just name this `github-metrics-aggregator` or `ghma`?

File: pkg/cli/retry.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		return err
+	}
+	defer closer()
+
+	return server.StartHTTPHandler(ctx, mux)
+}
+
+func (c *RetryServerCommand) RunUnstarted(ctx context.Context, args []string) (*serving.Server, http.Handler, func(), error) {
+	closer := func() {}
```
Comment (sethvargo) [inline pkg/cli/retry.go:68]: It doesn't look like we actually need to close anything?

File: pkg/cli/webhook.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return err
+	}
+	defer closer()
+
+	return server.StartHTTPHandler(ctx, mux)
+}
+
+func (c *WebhookServerCommand) RunUnstarted(ctx context.Context, args []string) (*serving.Server, http.Handler, func(), error) {
+	closer := func() {}
```
Comment (sethvargo) [inline pkg/cli/webhook.go:69]: Same here - we don't actually seem to be closing anything?

File: pkg/retry/server.go [hunk 4]
```diff
@@ -58,3 +67,10 @@ func (s *Server) handleRetry() http.Handler {
 		fmt.Fprint(w, `TODO: Make GitHub API calls and other stuff.. \n`)
 	})
 }
+
+func (s *Server) handleHealth() http.Handler {
```
Comment (sethvargo) [inline pkg/retry/server.go:71]: If you want a mini side project, we should move this (+ the grcp health) into the pkg/serving package to centralize it.

Comment (shan4993) [reply pkg/retry/server.go:71]: ack

File: pkg/cli/retry.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+`
+}
+
+func (c *RetryServerCommand) Flags() *cli.FlagSet {
+	c.cfg = &retry.Config{}
+	set := cli.NewFlagSet(c.testFlagSetOpts...)
+	return c.cfg.ToFlags(set)
+}
+
+func (c *RetryServerCommand) Run(ctx context.Context, args []string) error {
```
Comment (capri-xiyue) [inline pkg/cli/retry.go:57]: Do you need to add a health check for the retry server?

File: pkg/cli/retry_test.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+				if err := srv.StartHTTPHandler(serverCtx, mux); err != nil {
+					t.Error(err)
+				}
+			}()
+
+			client := &http.Client{
+				Timeout: 5 * time.Second,
+			}
+
+			uri := "http://" + srv.Addr() + "/healthz"
```
Comment (capri-xiyue) [inline pkg/cli/retry_test.go:139]: Where did you add the health check endpoint in the src?

File: pkg/retry/config.go [hunk 7]
```diff
@@ -79,3 +80,76 @@ func newConfig(ctx context.Context, lu envconfig.Lookuper) (*Config, error) {
...
+		Name:   "dataset-id",
+		Target: &cfg.DatasetID,
+		EnvVar: "DATASET_ID",
+		Usage:  `The dataset ID within the BigQuery instance.`,
+	})
+
+	f.DurationVar(&cli.DurationVar{
+		Name:    "lock-ttl-clock-skew",
+		Target:  &cfg.LockTTLClockSkew,
+		EnvVar:  "LOCK_TTL_CLOCK_SKEW",
```
Comment (capri-xiyue) [inline pkg/retry/config.go:126]: Add gma prefix to such config variables?

File: pkg/webhook/config.go [hunk 8]
```diff
@@ -82,3 +83,74 @@ func newConfig(ctx context.Context, lu envconfig.Lookuper) (*Config, error) {
...
+
+	f.StringVar(&cli.StringVar{
+		Name:   "project-id",
+		Target: &cfg.ProjectID,
+		EnvVar: "PROJECT_ID",
+		Usage:  `Google Cloud project ID.`,
+	})
+
+	f.IntVar(&cli.IntVar{
+		Name:   "retry-limit",
```
Comment (capri-xiyue) [inline pkg/webhook/config.go:135]: Same above, add gma prefix

File: pkg/retry/config.go (hunk 7 above)
Comment (shan4993) [reply pkg/retry/config.go:126]: its a little redundant so i left it off for GMA.

File: pkg/retry/server.go (hunk 4 above)
Comment (shan4993) [reply pkg/retry/server.go:71]: merged https://github.com/abcxyz/pkg/pull/110

File: .github/workflows/ci.yml (hunk 1 above)
Comment (shan4993) [reply .github/workflows/ci.yml:68]: ack was looking to be consistent but i agree, going with the former.

**Comment by shan4993** [comment]
FYI build is still breaking until we get infra sorted.

**Comment by sethvargo** [comment]
@shan4993 looks like you need to pass "withoutcredentials" to some of the tests so that it doesn't try to find creds in unit tests.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -125,10 +127,35 @@ jobs:
...
+        uses: 'google-github-actions/setup-gcloud@d51b5346f85640ec2aa2fa057354d2b82c2fcbce' # ratchet:google-github-actions/setup-gcloud@v1
+      - name: 'Deploy to Cloud Run'
+        run: |-
+          gcloud run services update ${{ env.INTEGRATION_RETRY_SERVICE_NAME }} \
+            --project="${{ env.INTEGRATION_RETRY_PROJECT_ID }}" \
+            --region="${{ env.INTEGRATION_RETRY_REGION }}" \
+            --image="${{ env.DOCKER_REPO }}/github-metrics-aggregator:${{ env.DOCKER_TAG }}-amd64"            
 
   integration:
     runs-on: 'ubuntu-latest'
```
[the remark this answers is not in the document]
Comment (shan4993) [reply .github/workflows/ci.yml:161]: ack

**Pull request merged**